libs/full/runtime_local/src/runtime_local.cpp

```diff
@@ -329,39 +319,55 @@ namespace hpx {
       , on_stop_func_(global_on_stop_func)
       , on_error_func_(global_on_error_func)
       , result_(0)
-      , main_pool_notifier_(std::move(main_pool_notifier))
-      , main_pool_(1, main_pool_notifier_, "main_pool")
+      , main_pool_notifier_()
+      , main_pool_(main_pool_notifier_, "main_pool")
 #ifdef HPX_HAVE_IO_POOL
-      , io_pool_notifier_(std::move(io_pool_notifier))
-      , io_pool_(rtcfg_.get_thread_pool_size("io_pool"), io_pool_notifier_,
-            "io_pool")
+      , io_pool_notifier_()
+      , io_pool_(io_pool_notifier_, "io_pool")
 #endif
 #ifdef HPX_HAVE_TIMER_POOL
-      , timer_pool_notifier_(std::move(timer_pool_notifier))
-      , timer_pool_(rtcfg_.get_thread_pool_size("timer_pool"),
-            timer_pool_notifier_, "timer_pool")
+      , timer_pool_notifier_()
+      , timer_pool_(timer_pool_notifier_, "timer_pool")
 #endif
-      , notifier_(std::move(notifier))
-      , thread_manager_(new hpx::threads::threadmanager(rtcfg_,
-#ifdef HPX_HAVE_TIMER_POOL
-            timer_pool_,
-#endif
-            notifier_, network_background_callback))
+      , notifier_()
+      , thread_manager_()
       , stop_called_(false)
       , stop_done_(false)
     {
-        // This needs to happen as early as possible to set up the runtime
-        // pointer.
         LPROGRESS_;
+    }
 
-        // initialize our TSS
+    void runtime::set_notification_policies(notification_policy_type&& notifier,
+#ifdef HPX_HAVE_IO_POOL
+        notification_policy_type&& io_pool_notifier,
+#endif
+#ifdef HPX_HAVE_TIMER_POOL
+        notification_policy_type&& timer_pool_notifier,
+#endif
+        threads::detail::network_background_callback_type
+            network_background_callback)
+    {
+        notifier_ = std::move(notifier);
+
+        main_pool_.init(1);
+#ifdef HPX_HAVE_IO_POOL
+        io_pool_notifier_ = std::move(io_pool_notifier);
+        io_pool_.init(rtcfg_.get_thread_pool_size("io_pool"));
+#endif
+#ifdef HPX_HAVE_TIMER_POOL
+        timer_pool_notifier_ = std::move(timer_pool_notifier);
+        timer_pool_.init(rtcfg_.get_thread_pool_size("timer_pool"));
+#endif
+
+        thread_manager_.reset(new hpx::threads::threadmanager(rtcfg_,
+#ifdef HPX_HAVE_TIMER_POOL
+            timer_pool_,
+#endif
+            notifier_, network_background_callback));
+
+        // initialize our TLS
         runtime::init_tss();
         util::reinit_construct();    // call only after TLS was initialized
```

Comment:
`init_tss` and `reinit_construct` don't seem to belong in `set_notification_policies`? Can they be moved to after the call to `set_notification_policies`?

Reply:
Yes, definitely.